framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -993,32 +991,48 @@ private void checkContractsAtMethodDeclaration(
         if (contracts.isEmpty()) {
             return;
         }
-
-        JavaExpressionContext jeContext =
-                JavaExpressionContext.buildContextForMethodDeclaration(methodTree, checker);
-
+        StringToJavaExpression stringToJavaExpr =
+                stringExpr -> StringToJavaExpression.atMethodBody(stringExpr, methodTree, checker);
         for (Contract contract : contracts) {
             String expressionString = contract.expressionString;
             AnnotationMirror annotation = contract.annotation;
-
             annotation =
                     atypeFactory
                             .getDependentTypesHelper()
-                            .viewpointAdaptQualifierFromContract(annotation, jeContext, methodTree);
+                            .viewpointAdaptQualifierFromContract(
+                                    annotation, stringToJavaExpr, methodTree);
 
             JavaExpression exprJe;
             try {
-                exprJe = JavaExpressionParseUtil.parse(expressionString, jeContext);
+                // Parse the expressionString that was written on methodElement at the methodDecl.
+                exprJe = StringToJavaExpression.atMethodBody(expressionString, methodTree, checker);
             } catch (JavaExpressionParseException e) {
-                exprJe = null;
                 checker.report(methodTree, e.getDiagMessage());
+                if (formalParamNames != null && formalParamNames.contains(expressionString)) {
```

Comment:
I think there should be an error if the contract expression contains any formal parameter name as a subexpression, not just if the contract expression is a formal parameter name.

Reply:
See https://github.com/typetools/checker-framework/pull/4396.